### ISSUE TITLE
contentFor does not require this

### ISSUE THREAD

**Issue by cah-briangantzler** [cibernox/ember-power-select#220]
This is just an observation I learned when playing. I believe it is correct

I have a set of components in an addon (private for my company) that I am trying to use ember-power-select. When including my addon in my project I found an error was being thrown (ember-basic-dropdown-wormhole wasnt in the dom) yet this didnt happen my addons dummy app. A lot of research found that ember-power-selects blueprint was not adding ember-basic-dropdown as a package, but instead using the one in ember-power-selects node_modules due to a dependance. Taking that to heart I tried to do the same in my addon and found that doing so ember-power-selects contentFor did not work because the this referred to my addon not ember-power-select. I had to write the following code to work

```
    contentFor: function(type, config) {
        var emberPowerSelect = this.addons.filter(function(addon) {
            return addon.name === 'ember-power-select';
        })[0];
        return emberPowerSelect.contentFor(type, config);
    }
```

I am suspecting that the same code would work for ember-power-select. You do not need to call the apply with this. (Looking at ember-basic-dropdown it never uses this which is why it works at this time)

Also you will note there is a missing semi-colon after the [0] in the this addon.

If I then add ember-power-select to my project to use in other places, I note that ember-basic-dropdown-wormhole <div> gets added twice with the same id (thought that wasnt allowed) but all still works since it uses the first found.

So as a question from a learning perspective. Should addons be adding the other addons they depend on to the project with an addPackagesToProject or attempting to use the embedded addon? 

Proxying contentFor could lead to the contentFor being called multiple times and quite possibly on different versions of the addon.

**Comment by cibernox** [comment]
Hey, sorry for the delay. Xmas holidays and all that....

Regarding the `this` when invoking the other addon, is that wrong? I know that is not requires, but iirc, when you do `super` in ES6 code or `this._super(...arguments)`, that transpiles to `myFnInThePrototype.apply(this, arguments)`, and so did it. You mention it just because it not needed or because it has any undesired consecuence.

Actually I'm not entirely sure what is the proper way of handing 2nd degree dependencies.

I see your point that if installing ember-power-select adds ember-basic-dropdown, but then the user adds ember-basic-dropdown on its own then the div will be duplicated.
However, ideally the fact that ember-power-select uses ember-basic-dropdown should be an implementation detail. Only your direct dependencies should be listed in the packages.json.

I probably should ask the wise elders in slack about this.

**Comment by cibernox** [comment]
@cah-briangantzler I've been asking in the slack channel what is the good approach for this and seems that people favor not adding dependencies using the `addPackagesToProject`.

I've released 0.8.0-beta.2 that depends on ember-basic-dropdown ~0.7.2. 
That new version of ember-basic-dropdown has an idempotent `contentFor` hook that won't do anything if it's the second time it is invoked. This remove the duplicated destination div.

**Comment by cah-briangantzler** [comment]
Thanks for asking, I was wondering what to do as well. And I like the approach of only direct dependencies.

However, I would hope all the hooks for ALL addons would be called by ember-cli and not just directs, and then addons have to proxy down to dependent addons.

So bascially contentFor (in this case) is a hook for an addon, not overriding contentFor for all your dependent addons. Was the contentFor for ember-basic-dropdown NOT being called and that is why you are calling it? I would ask about that in the slack channels. The behaviour I would expect from CLI would be to resolve all the dependancies for a list of addons, then initialize them wether they are direct or not.

Say two addons depend on ember-basic-dropdown and one depends on version A and one depends on version B. When you do an NPM install it lists the multiple dependencies and you choose which dependency wins. That dependencies hooks should be called by ember CLI not have the addons that depended on it to delegate its hooks onto depended addon hooks. Is that making sense?

I created an addon that uses power-select, so I did exactly what you did, index.js filtering addons and calling the contentFor for power-select and what happens was when this.addons.filter executed in power-select it filtered through MY addons not power-select addons cause "this" was my addon, not power select. So currently there is no negative effect because ember-basic-dropdown does NOT refer to "this" in its contentFor, but if it did, this will point to ember-power-select not ember-basic-dropdown so there is a potential future bug.
